mirrord/cli/src/wizard.rs

```diff
@@ -0,0 +1,671 @@
+/// need to do it manually if making changes to the wizard.
+const COMPRESSED_FRONTEND: &[u8] = include_bytes!("../../../wizard-frontend.tar.gz");
+
+/// The entrypoint for the `wizard` command. Unzips the frontend and serves it using axum, along
+/// with internal API endpoints.
+pub async fn wizard_command<P>(
+    user_data: UserData,
+    args: WizardArgs,
+    parent_progress: &mut P,
+) -> CliResult<()>
+where
+    P: Progress,
+{
+    let mut progress = parent_progress.subtask("launching wizard");
+    let is_returning_string = user_data.is_returning_wizard().to_string();
+
+    // client for fetching targets from the cluster when detecting exposed ports
+    let client = create_kube_config(
+        args.accept_invalid_certificates,
+        args.kubeconfig,
+        args.context,
+    )
+    .await
+    .and_then(|config| Ok(ClientBuilder::try_from(config.clone())?.build()))
+    .map_err(|error| CliError::friendlier_error_or_else(error, CliError::CreateKubeApiFailed))?;
+
+    // unpack the frontend into `TempDir`
+    let tmp_dir = TempDir::new()?;
+    let temp_dir_path = tmp_dir.path();
+
+    let tar_gz = Cursor::new(COMPRESSED_FRONTEND);
+    let tar = GzDecoder::new(tar_gz);
+    let mut archive = Archive::new(tar);
+    archive.unpack(temp_dir_path)?;
+
+    let serve_client = {
+        let index_service = ServiceBuilder::new()
+            .layer(SetResponseHeaderLayer::overriding(
+                header::CACHE_CONTROL,
+                header::HeaderValue::from_static("no-cache, no-store, must-revalidate"),
+            ))
+            .layer(SetResponseHeaderLayer::overriding(
+                header::PRAGMA,
+                header::HeaderValue::from_static("no-cache"),
+            ))
+            .layer(SetResponseHeaderLayer::overriding(
+                header::EXPIRES,
+                header::HeaderValue::from_static("0"),
+            ))
+            .service(ServeFile::new(temp_dir_path.join("index.html")));
+
+        ServiceBuilder::new()
+            .layer(SetResponseHeaderLayer::if_not_present(
+                header::CACHE_CONTROL,
+                header::HeaderValue::from_static("public, max-age=1209600"),
+            ))
+            .service(
+                ServeDir::new(temp_dir_path)
+                    .fallback(index_service)
+                    .append_index_html_on_directories(false),
+            )
+    };
+
+    // prepare temp dir as SPA with endpoints behind `/api` path
+    let app = Router::new()
+        .layer(TraceLayer::new_for_http())
+        .fallback_service(serve_client)
+        .route(
+            "/api/v1/is-returning",
+            get(|| async { is_returning_string }),
+        )
+        .route("/api/v1/cluster-details", get(cluster_details))
+        .route("/api/v1/namespace/{namespace}/targets", get(list_targets))
+        .with_state(Arc::new(Mutex::new((user_data, client))));
```

Comment:
Probably better to create a `BackendState` struct. Otherwise it'll get very annoying to make changes to it.